### PR TITLE
Diamond

### PR DESCRIPTION
**Pull request by michalsoloducha**

### CODE REVIEW

**Comment by ziobron** [comment]
✅ CI Passed - Task: cars-tests
🏅 Granted 10 XP for @michalsoloducha

**Comment by ziobron** [comment]
✅ CI Passed - Task: cars-exception
🏅 Granted 5 XP for @michalsoloducha

File: Car.hpp [hunk 1]
```diff
@@ -1,12 +1,21 @@
...
     void turnRight();
     void brake();
-    void accelerate(int speed);
+    void setSpeed(const int);
+    size_t getSpeed();
     virtual void refill() = 0;
-    virtual ~Car(){}
+    virtual void changeEngine(std::unique_ptr<PetrolEngine>);
+    virtual void changeEngine(std::unique_ptr<ElectricEngine>);
+    virtual ~Car() {}
```
Comment (nauka-programowania-MA) [inline Car.hpp:17]: = default. And remember about the rule of 5!

File: Car.hpp [hunk 2]
```diff
@@ -1,12 +1,21 @@
...
+class ElectricEngine;
+
+class Car {
 public:
     void turnLeft();
     void turnRight();
     void brake();
-    void accelerate(int speed);
+    void setSpeed(const int);
+    size_t getSpeed();
```
Comment (nauka-programowania-MA) [inline Car.hpp:13]: const

File: ElectricCar.hpp [hunk 3]
```diff
@@ -1,17 +1,16 @@
...
 #include "Car.hpp"
 #include "ElectricEngine.hpp"
 
-class ElectricCar : public virtual Car
-{
+class ElectricCar : public virtual Car {
 public:
-    ElectricCar(ElectricEngine* engine);
+    ElectricCar(std::unique_ptr<ElectricEngine>);
     ~ElectricCar();
```
Comment (nauka-programowania-MA) [inline ElectricCar.hpp:9]: rule of 5

File: HybridCar.hpp [hunk 4]
```diff
@@ -1,14 +1,13 @@
...
 #include "PetrolEngine.hpp"
-#include <iostream>
 
-class HybridCar : public ElectricCar, public PetrolCar
-{
+class HybridCar : public ElectricCar, public PetrolCar {
 public:
-    HybridCar(PetrolEngine* petrolEng, ElectricEngine* electricEng);
+    HybridCar(std::unique_ptr<PetrolEngine>, std::unique_ptr<ElectricEngine>);
     ~HybridCar();
```
Comment (nauka-programowania-MA) [inline HybridCar.hpp:11]: rule of 5

File: PetrolCar.hpp [hunk 5]
```diff
@@ -1,18 +1,18 @@
...
+#include "PetrolEngine.hpp"
 
-
-class PetrolCar : virtual public Car
-{
+class PetrolCar : virtual public Car {
 public:
-    PetrolCar(PetrolEngine* engine);
+    PetrolCar(std::unique_ptr<PetrolEngine>);
     ~PetrolCar();
```
Comment (nauka-programowania-MA) [inline PetrolCar.hpp:9]: rule of 5

**Review by nauka-programowania-MA — APPROVED** [review]

**Pull request closed without merging**